Fixed Media Layout in RTL, fix #2101.

huangyz0918 opened the pull request:
Closes #2101 

#### What has been done to verify that this works as intended?
Despite of API 16, I made some changes in `MediaLayout.java` and make it mirrored from LTR.

Screen Records in different APIs:

| Device API | LTR  | RTL |
| :------------: |:---------------:|:-----:|
| Android 4.4      | <img src="https://i.loli.net/2018/04/12/5acf13599e5d7.png" width = "200" align=center /> | <img src="https://i.loli.net/2018/04/12/5acf134405a36.png" width = "200" align=center /> |
| Android 5.0      | <img src="https://i.loli.net/2018/04/12/5acf137b8a982.png" width = "200" align=center /> | <img src="https://i.loli.net/2018/04/12/5acf1368af9df.png" width = "200" align=center /> |
| Android 6.0      | <img src="https://i.loli.net/2018/04/12/5acf13db419ac.png" width = "200" align=center /> | <img src="https://i.loli.net/2018/04/12/5acf138d0fdc1.png" width = "200" align=center /> |
| Android 7.1      | <img src="https://i.loli.net/2018/04/12/5acf13fcdf6e7.png" width = "200" align=center /> | <img src="https://i.loli.net/2018/04/12/5acf13ed0adee.png" width = "200" align=center /> |
| Android 8.1      | <img src="https://i.loli.net/2018/04/12/5acf140b93ff0.png" width = "200" align=center /> | <img src="https://i.loli.net/2018/04/12/5acf141b5ee1a.png" width = "200" align=center /> |

#### Why is this the best possible solution? Were any other approaches considered?
It's the best solution without a refactor but not the most readable code, if any one have spare time, welcome to refactor that ! 👍 ——> issue #2127 
#### Are there any risks to merging this code? If so, what are they?
No.
#### Do we need any specific form for testing your changes? If so, please attach one.
Recommend `Birds.xml` or something else.
#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.
No.

huangyz0918 commented:
Known issues：

1. Cannot cover **API level 16** ( I just skipped  that ).

2. Android 4.2 ( API level 17 ) **has some unnecessary margins** in media buttons.

huangyz0918 commented:
@grzesiek2010 @lognaturel Please review if you have spare time.
Thank you ☺️

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java:
```diff
@@ -362,35 +359,30 @@ public void onClick(View v) {
             // Text upper left; audio upper right; video below audio on right.
             // image below text, audio and video buttons; left-aligned with text.
-            textParams.addRule(RelativeLayout.ALIGN_PARENT_LEFT);
+            if (isNeedPatchRTL()) {
```
̶I̶ ̶d̶o̶n̶'̶t̶ ̶u̶n̶d̶e̶r̶s̶t̶a̶n̶d̶ ̶t̶h̶i̶s̶ ̶f̶r̶a̶g̶m̶e̶n̶t̶ ̶`̶A̶L̶I̶G̶N̶_̶P̶A̶R̶E̶N̶T̶_̶S̶T̶A̶R̶T̶`̶ ̶a̶n̶d̶ ̶`̶A̶L̶I̶G̶N̶_̶P̶A̶R̶E̶N̶T̶_̶L̶E̶F̶T̶`̶ ̶s̶h̶o̶u̶l̶d̶ ̶b̶e̶ ̶t̶h̶e̶ ̶s̶a̶m̶e̶ ̶s̶o̶ ̶t̶h̶e̶y̶ ̶a̶r̶e̶ ̶u̶s̶e̶d̶ ̶u̶n̶d̶e̶r̶ ̶d̶i̶f̶f̶e̶r̶e̶n̶t̶ ̶c̶i̶r̶c̶u̶m̶s̶t̶a̶n̶c̶e̶s̶?̶

huangyz0918 replied:
I'm sorry, do you mean using `ALIGN_PARENT_START` and `ALIGN_PARENT_LEFT` will lead to the same result? (in both RTL and LTR)

grzesiek2010 replied:
No, I just had a temporary black hole in my head and didn't catch it.

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java:
```diff
@@ -411,6 +403,40 @@ public void onClick(View v) {
+                                     RelativeLayout.LayoutParams imageParams,
+                                     int imagePadding) {
+        mediaParams.addRule(RelativeLayout.ALIGN_PARENT_TOP);
+        if (isNeedPatchRTL()) {
```
̶T̶h̶e̶ ̶s̶a̶m̶e̶ ̶a̶s̶ ̶a̶b̶o̶v̶e̶ ̶b̶u̶t̶ ̶h̶e̶r̶e̶ ̶w̶i̶t̶h̶ ̶`̶A̶L̶I̶G̶N̶_̶P̶A̶R̶E̶N̶T̶_̶E̶N̶D̶`̶ ̶a̶n̶d̶ ̶`̶A̶L̶I̶G̶N̶_̶P̶A̶R̶E̶N̶T̶_̶R̶I̶G̶H̶T̶`̶

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java:
```diff
@@ -411,6 +403,40 @@ public void onClick(View v) {
+
+    // used for items without image
+    private void injectRulesByLocale(RelativeLayout.LayoutParams mediaParams) {
+        if (isNeedPatchRTL()) {
```
̶A̶n̶d̶ ̶o̶n̶e̶ ̶m̶o̶r̶e̶ ̶t̶i̶m̶e̶.̶

grzesiek2010 commented on collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java:
```diff
@@ -302,6 +304,9 @@ public void onClick(View v) {
 
         // Determine the layout constraints...
         // Assumes LTR, TTB reading bias!
+        int startOf = isNeedPatchRTL() ? RelativeLayout.ALIGN_PARENT_START : RelativeLayout.ALIGN_PARENT_LEFT;
```
̶a̶n̶d̶ ̶t̶h̶e̶ ̶s̶a̶m̶e̶ ̶h̶e̶r̶e̶.̶

grzesiek2010 commented:
Ok, you can ignore my comments.
Left is always left but strat is left or right depending on the language of course.

huangyz0918 committed:
🐛 fix: Fixed media layouts in RTL

huangyz0918 committed:
✏️ typo: Fixed typo in comments

lognaturel commented on collect_app/src/main/java/org/odk/collect/android/views/MediaLayout.java:
```diff
@@ -302,6 +304,8 @@ public void onClick(View v) {
 
         // Determine the layout constraints...
         // Assumes LTR, TTB reading bias!
```
This comment is no longer accurate! How about changing it to "Determine the layout constraints. Right-to-left support works for API > 16."

huangyz0918 committed:
💬 texts: updated comments

huangyz0918 commented:
@opendatakit-bot unlabel "needs review"

mmarciniak90 commented:
@huangyz0918, @lognaturel I do not feel like a specialist on RTL languages.
I agree that view with media is mirrored from LTR.

|master | fix LTR| fix RTL| 
|---|---|---|
|![screenshot_20180521-105412](https://user-images.githubusercontent.com/3907554/40299692-9cf2eb80-5cd6-11e8-8728-cdf3b3be541c.png)|![screenshot_20180521-115114](https://user-images.githubusercontent.com/3907554/40299702-a2e0e7cc-5cd6-11e8-8e5c-b5a2040f91df.png)|![screenshot_20180521-121045](https://user-images.githubusercontent.com/3907554/40299794-f1973ca4-5cd6-11e8-85ef-6385f2b29a5f.png)|

I'm not sure that this fix should cause changes on this view too:

|master |fix |
|---|---|
|![screenshot_20180521-105240](https://user-images.githubusercontent.com/3907554/40299907-47eea74a-5cd7-11e8-992c-e22b58891fda.png)|![screenshot_20180521-115054](https://user-images.githubusercontent.com/3907554/40299913-4ba0eefc-5cd7-11e8-85f2-616c46be7bf2.png)|

* Group brackets are incorect on Android 4.4 and 4.2
* First question is not fully visible. Speaker icon covers question mark.
![screenshot_2018-05-21-12-37-02](https://user-images.githubusercontent.com/3907554/40301624-23dd21a6-5cdc-11e8-9092-6b714f1b1608.png)
* No changes on Android 4.1
![screenshot_2018-05-21-11-47-59](https://user-images.githubusercontent.com/3907554/40301848-d9715d52-5cdc-11e8-81e5-d0c3b9662e93.png)

huangyz0918 commented:
> I'm not sure that this fix should cause changes on this view too:

I can fix this.

> Group brackets are incorect on Android 4.4 and 4.2

And this (but I'm not sure this is a common issue since I cannot reproduce it in Android 4.4, you can see in my screenshot).

> No changes on Android 4.1

We just gave up API 16 since it's not supported by RTL.

huangyz0918 committed:
🐛 fix: fixed textviews

huangyz0918 committed:
🚜 ci: fixed pmd

huangyz0918 committed:
🚑 quickfix: fixed the brackets in 4.4 && 4.2

huangyz0918 commented:
@mmarciniak90 I made some improvements.
> I'm not sure that this fix should cause changes on this view too:

I made this view mirrored from LTR too. (> android 4.2)

> Group brackets are incorect on Android 4.4 and 4.2

Fixed. 

> First question is not fully visible. Speaker icon covers question mark.

Sorry, but I cannot reproduce that (I adjusted the margins and hope it works).

mmarciniak90 commented:
**Tested with success**

Verified on Android: 4.2, 4.4, 5.1, 6.0, 7.0, 8.1
Mentioned problems were fixed. @huangyz0918 great job! :1st_place_medal: 
Verified on light and dark mode. 

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"
